IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,539 @@
+    MSID_LOG_INFO_PII(context, @"Set keychain item, key info (account: %@ service: %@)", key.account, key.service);
+
+    if (!key.service)
+    {
+        NSString *errorMessage = @"Set keychain item with invalid key (service is nil).";
+        MSID_LOG_WARN(context, @"%@", errorMessage);
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, (NSInteger)MSIDErrorInternal, errorMessage, nil, nil, nil, context.correlationId, nil);
+        }
+        return FALSE;
+    }
+
+    NSData *jsonData = [serializer serializeAccountCacheItem:account];
+    if (!jsonData)
+    {
+        NSString *errorMessage = @"Failed to serialize account to json data.";
+        MSID_LOG_WARN(context, @"%@", errorMessage);
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, (NSInteger)MSIDErrorInternal, errorMessage, nil, nil, nil, context.correlationId, nil);
+        }
+        return FALSE;
+    }
+
+    NSMutableDictionary *query = [self defaultAccountQuery:key];
+    NSMutableDictionary *update = [self defaultAccountUpdate:key];
+    update[(id)kSecValueData] = jsonData;
+    OSStatus status = SecItemUpdate((CFDictionaryRef)query, (CFDictionaryRef)update);
+    MSID_LOG_INFO(context, @"Keychain update status: %d", (int)status);
+
+    if (status == errSecItemNotFound)
+    {
+        [query addEntriesFromDictionary:update];
+        status = SecItemAdd((CFDictionaryRef)query, NULL);
+        MSID_LOG_INFO(context, @"Keychain add status: %d", (int)status);
+    }
+
+    if (status != errSecSuccess)
+    {
+        NSString *errorMessage = @"Failed to write account to keychain";
+        MSID_LOG_WARN(context, @"%@ (%d)", errorMessage, status);
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDKeychainErrorDomain, (NSInteger)status, errorMessage, nil, nil, nil, context.correlationId, nil);
+        }
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+// Read a single account from the macOS keychain cache.
+// If multiple matches are found, return nil and set an error.
+//
+// Errors:
+// * MSIDErrorDomain/MSIDErrorCacheMultipleUsers: more than one keychain item matched the account key
+//
+- (MSIDAccountCacheItem *)accountWithKey:(MSIDCacheKey *)key
+                              serializer:(id<MSIDAccountItemSerializer>)serializer
+                                 context:(id<MSIDRequestContext>)context
+                                   error:(NSError **)error
+{
+    MSID_TRACE;
+    NSArray<MSIDAccountCacheItem *> *items = [self accountsWithKey:key
+                                                        serializer:serializer
+                                                           context:context
+                                                             error:error];
+
+    if (items.count > 1)
+    {
+        if (error)
+        {
+            NSString *errorMessage = @"The token cache store for this resource contains more than one user";
```

Comment:
what resource?

Reply:
This wording comes from the iOS implementation, but it's referring to the search key.